### DISCUSSION TITLE
brew upgrade: Could not kickstart service "com.apple.audio.coreaudiod": 1: Operation not permitted

### DISCUSSION THREAD

**Discussion by beremour** [ExistentialAudio/BlackHole#781] [General]
Upgrading blackhole-16ch 0.5.0 -> 0.6.0 gives error:

```brew -v upgrade
==> Casks with 'auto_updates true' or 'version :latest' will not be upgraded; pass `--greedy` to upgrade them.
==> Upgrading 1 outdated package:
blackhole-16ch 0.5.0 -> 0.6.0
==> Upgrading blackhole-16ch
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-cask/ef939c438c5519c1dc533f8384691106bc2262d9/Casks/b/blackhole-16ch.rb
Already downloaded: /Users/ruth/Library/Caches/Homebrew/downloads/849d5b814b750eda8a8a7f27610d817689b8b36f08e0541805107c61109efc56--blackhole-16ch.rb
==> Verifying checksum for '849d5b814b750eda8a8a7f27610d817689b8b36f08e0541805107c61109efc56--blackhole-16ch.rb'
==> Downloading https://existential.audio/downloads/BlackHole16ch-0.6.0.pkg
Already downloaded: /Users/ruth/Library/Caches/Homebrew/downloads/c01900038e21221d409bcb4db5885706da825cc15f7194af3c902624f016d668--BlackHole16ch-0.6.0.pkg
==> Verifying checksum for 'c01900038e21221d409bcb4db5885706da825cc15f7194af3c902624f016d668--BlackHole16ch-0.6.0.pkg'
==> Uninstalling packages with sudo; the password may be necessary:
Could not kickstart service "com.apple.audio.coreaudiod": 1: Operation not permitted
==> Purging files for version 0.6.0 of Cask blackhole-16ch
Error: blackhole-16ch: Failure while executing; `/usr/bin/sudo -E -- /bin/launchctl kickstart -kp system/com.apple.audio.coreaudiod` exited with 1. Here's the output:
Could not kickstart service "com.apple.audio.coreaudiod": 1: Operation not permitted```

**Comment by beremour** [comment]
Forgot to mention: it's Sonoma 14.4.1 on M3 chip.

**Comment by Zakharchos** [comment]
**This helped on my M1**
https://github.com/ExistentialAudio/BlackHole/issues/779#issuecomment-2058581435
> I found a way to workaround this:
> 
> You go to the Caskroom for the cask, on Apple Silicon macs, it is here:
> 
> ```
> /opt/homebrew/Caskroom/blackhole-2ch/.metadata/0.5.0/
> ```
> 
> On Intel Macs, it is here:
> 
> ```
> /usr/local/Caskroom/blackhole-2ch/.metadata/0.5.0/
> ```
> 
> And you will see a folder with a date, cd into it, you will see the `blackhole-2ch.rb`
> 
> Edit it that file, change
> 
> ```ruby
> system_command "/bin/launchctl",
>                    args:         [
>                      "kickstart",
>                      "-kp",
>                      "system/com.apple.audio.coreaudiod",
>                    ],
>                    sudo:         true,
>                    must_succeed: true
> ```
> 
> to
> 
> ```ruby
> system_command "/usr/bin/killall",
>                    args:         ["coreaudiod"],
>                    sudo:         true,
>                    must_succeed: true
> ```
> 
> Then save the file, and then uninstall should success.


After editing the `.rb` just do a `brew uninstall` and then a `brew install`

**Reply by 101Dude** [reply]
this^^

**Reply by miranquil** [reply]
superb!

**Reply by asbjornu** [reply]
This worked for [background-music](https://formulae.brew.sh/cask/background-music) as well, which failed with the same error. The path to the formula was `/usr/local/Caskroom/background-music/.metadata/0.4.0/20221019085700.540/Casks/background-music.rb` for me and changing the `uninstall_postflight` block to the above, made it possible to uninstall.

**Reply by kaovilai** [reply]
```
❯ find -L /usr/local/ -name 'background-music.rb'
/usr/local//homebrew/Library/Taps/homebrew/homebrew-cask/Casks/background-music.rb
```
Mine is in a different location.

**Reply by Mitlanyal** [reply]
> ```ruby
> system_command "/usr/bin/killall",
>                    args:         ["coreaudiod"],
>                    sudo:         true,
>                    must_succeed: true
> ```

perfection, thank you.

**Comment by beremour** [comment]
I've got it on
```/opt/homebrew/Caskroom/blackhole-16ch/.metadata/0.5.0/20240219133217.677/Casks/blackhole-16ch.rb```
It helped, thank you very much!

**Comment by flechtwerkflechtwerk** [comment]
Workaround worked for me too! Thanks.

**Comment by thecsw** [comment]
workaround worked for me too! thanks

**Comment by lox** [comment]
FYI, 14.4 added a "security" patch that disables the use of kickstart to restart coreaudio:

> To improve security and stability, using launchctl kickstart -k is no longer permitted for critical system processes. If a process must be forcefully terminated, it is recommended to use kill instead. (123028502)

**Comment by h2ogeek** [comment]
Thank you @Zakharchos I had the same issue with `background-music` and your suggestion took care of this issue :-)

**Comment by limonkufu** [comment]
Thanks! I had the same error for `background-music` and workaround [here](https://github.com/ExistentialAudio/BlackHole/discussions/781#discussioncomment-9144175) works

**Comment by headhunter45** [comment]
I'm just adding because this is the first result on bing for this error message when I searched. I bounce between different search engines a lot.

Here is the error on apple developer forums: https://forums.developer.apple.com/forums/thread/748228
and here is the fix from there: https://developer.apple.com/documentation/macos-release-notes/macos-14_4-release-notes#Core-Audio

Essentially Apple stopped allowing users to restart critical system services to protect system integrity and now they recommend killing the process instead. I'm not sure if we have to manually restart the service after or not. I can't test atm.

**Reply by daslicht** [reply]
does the latest version consider that change ?

**Comment by devinroth** [comment]
Yes
